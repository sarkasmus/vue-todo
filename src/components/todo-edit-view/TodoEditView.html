<div class="todo-edit">
  <header class="header text-center">
    <h1>{{ $route.name | capitalize }}</h1>
  </header>

  <!-- List of todos. -->
  <div class="form">
    <div class="form-group">
      <input class="prop-title form-control" v-model="title" placeholder="Title">
    </div>

    <div class="form-group">
      <textarea class="description form-control" placeholder="Description for a todo" v-model="description"></textarea>
    </div>

    <div class="form-group">
      <input type="date" class="dueDate form-control" v-model="dueDate" placeholder="Date">
    </div>
  </div>

  <footer>
    <button class="save btn btn-info" @click="save" :disabled="!(title.length > 3 && description.length > 5 && Date.parse(dueDate) > 0)">Save</button>
  </footer>
</div>
