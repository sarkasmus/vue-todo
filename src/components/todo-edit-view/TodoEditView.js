export default {
  name: 'TodoEditView',
  data () {
    // Load todo data.
    if (typeof this.$store.getters.todos[this.$route.params.id] !== 'undefined') {
      return {
        title: this.$store.getters.todos[this.$route.params.id].title,
        description: this.$store.getters.todos[this.$route.params.id].description,
        dueDate: this.$store.getters.todos[this.$route.params.id].dueDate
      }
    }

    return {
      title: '',
      description: '',
      dueDate: ''
    }
  },
  methods: {
    save () : void {
      if (this.$route.name === 'create') {

        // Create a new todo.
        this.$store.dispatch('create', {
          title: this.title,
          description: this.description,
          dueDate: this.dueDate
        })
      } else {

        // Edit todo of id.
        this.$store.dispatch('edit', {
          id: this.$route.params.id,
          title: this.title,
          description: this.description,
          dueDate: this.dueDate
        })
      }

      this.$router.push('/')
    }
  },
  filters: {
    capitalize (text: String) : String {
      if (!text) {
        return ''
      }

      text = text.toString()
      return text.charAt(0).toUpperCase() + text.slice(1)
    }
  }
}
