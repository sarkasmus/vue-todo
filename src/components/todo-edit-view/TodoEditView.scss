.form {
  margin-top: 50px;
}

.header h1 {
  font-size: 56px;
  color: #333333
}

.header h2 {
  font-size: 22px;
  font-weight: normal;
  color: #aaaaaa
}
