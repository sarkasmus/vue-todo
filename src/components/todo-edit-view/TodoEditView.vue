<template src="./TodoEditView.html"></template>
<style src="./TodoEditView.scss"></style>
<script src="./TodoEditView.js"></script>
