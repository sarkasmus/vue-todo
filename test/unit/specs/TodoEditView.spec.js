import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import TodoEditView from '../../../src/components/todo-edit-view/TodoEditView.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

/**
 * Testing component representing single task (todo).
 */
describe('Todo', () => {
  let todoEditViewComponent
  let actions
  let getters
  let router

  /**
   * Inits every testcase.
   */
  beforeEach(() => {
    actions = {
      create: jest.fn(),
      edit: jest.fn()
    }

    getters = {
      todos () {
        return [
          {
            title: 'Task 1',
            description: 'My awesome task',
            finished: false,
            dueDate: '2018-06-02'
          },
          {
            title: 'Task 2',
            description: 'My awesome task',
            finished: true,
            dueDate: '2018-06-03'
          }
        ]
      }
    }

    let store = new Vuex.Store({
      actions: actions,
      getters: getters
    })

    router = {
      push: jest.fn()
    }

    todoEditViewComponent = shallowMount(TodoEditView, {
      mocks: {
        $router: router,
        $route: {
          name: 'create',
          params: {}
        }
      },
      store,
      localVue
    })
  })

  it('should be able to create a new task', () => {
    todoEditViewComponent.find('.prop-title').element.value = 'Input'
    todoEditViewComponent.find('.prop-title').trigger('input')

    todoEditViewComponent.find('.description').element.value = 'Description'
    todoEditViewComponent.find('.description').trigger('input')

    todoEditViewComponent.find('.dueDate').element.value = '2018-06-04'
    todoEditViewComponent.find('.dueDate').trigger('input')

    todoEditViewComponent.find('.save').trigger('click')

    expect(actions.create).toBeCalled()
    expect(router.push).toBeCalled()
  })

  it('should be able to edit an existing task', () => {
    todoEditViewComponent.vm.$route = {
      name: 'edit',
      params: {
        id: 1
      }
    }

    todoEditViewComponent.find('.prop-title').element.value = 'Input'
    todoEditViewComponent.find('.prop-title').trigger('input')

    todoEditViewComponent.find('.description').element.value = 'Description'
    todoEditViewComponent.find('.description').trigger('input')

    todoEditViewComponent.find('.dueDate').element.value = '2018-06-04'
    todoEditViewComponent.find('.dueDate').trigger('input')

    todoEditViewComponent.find('.save').trigger('click')

    expect(actions.edit).toBeCalled()
    expect(router.push).toBeCalled()
  })

  /**
   * Matching snapshot - HTML structure of the component.
   */
  it('should match snapshot', () => {
    expect(todoEditViewComponent.element).toMatchSnapshot()
  })

  /**
   * Matching edit page.
   */
  it('should match snapshot', () => {
    todoEditViewComponent.vm.$route = {
      name: 'edit',
      params: {
        id: 1
      }
    }

    expect(todoEditViewComponent.element).toMatchSnapshot()
  })
})
