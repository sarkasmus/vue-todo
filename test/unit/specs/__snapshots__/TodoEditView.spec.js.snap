// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should match snapshot 1`] = `
<div
  class="todo-edit"
>
  <header
    class="header text-center"
  >
    <h1>
      Create
    </h1>
  </header>
   
  <div
    class="form"
  >
    <div
      class="form-group"
    >
      <input
        class="prop-title form-control"
        placeholder="Title"
      />
    </div>
     
    <div
      class="form-group"
    >
      <textarea
        class="description form-control"
        placeholder="Description for a todo"
      />
    </div>
     
    <div
      class="form-group"
    >
      <input
        class="dueDate form-control"
        placeholder="Date"
        type="date"
      />
    </div>
  </div>
   
  <footer>
    <button
      class="save btn btn-info"
      disabled="disabled"
    >
      Save
    </button>
  </footer>
</div>
`;

exports[`Todo should match snapshot 2`] = `
<div
  class="todo-edit"
>
  <header
    class="header text-center"
  >
    <h1>
      Create
    </h1>
  </header>
   
  <div
    class="form"
  >
    <div
      class="form-group"
    >
      <input
        class="prop-title form-control"
        placeholder="Title"
      />
    </div>
     
    <div
      class="form-group"
    >
      <textarea
        class="description form-control"
        placeholder="Description for a todo"
      />
    </div>
     
    <div
      class="form-group"
    >
      <input
        class="dueDate form-control"
        placeholder="Date"
        type="date"
      />
    </div>
  </div>
   
  <footer>
    <button
      class="save btn btn-info"
      disabled="disabled"
    >
      Save
    </button>
  </footer>
</div>
`;
